### PR TITLE
GH-3071 ensure bsa result iter returns result when assignment is UNDEF

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #3071 

Briefly describe the changes proposed in this PR:

- fix corner case of binding set assignment result processing when binding is UNDEF.
- added regresssion test
- some variable renaming and code cleanup 

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java [hunk 1]
```diff
@@ -730,52 +730,48 @@ protected Value getVarValue(Var var, BindingSet bindings) {
...
-					final BindingSet assignedBindings = iter.next();
+				QueryBindingSet nextResult = null;
+				while (nextResult == null && assignments.hasNext()) {
+					final BindingSet assignedBindings = assignments.next();
+
 					for (String name : assignedBindings.getBindingNames()) {
+						if (nextResult == null) {
+							nextResult = new QueryBindingSet(bindings);
+						}
+
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java:754]: This is the crux of the fix: we initialize `nextResult` as soon as we have a binding name (previously it was only initialized if that binding name also had a value in the BSA)

**Comment by hmottestad** [comment]
@jeenbroekstra tried to look at this several times but couldn’t really understand what’s going on. Sorry for my tardiness.

**Comment by jeenbroekstra** [comment]
> @jeenbroekstra tried to look at this several times but couldn’t really understand what’s going on. Sorry for my tardiness.

No worries, that's fine. For what it's worth as long as I know someone has taken a look and doesn't spot anything obviously wrong (even if they don't fully understand the entire fix), that is useful feedback to me. So don't feel you need to refrain from commenting even if you only have a few minutes to take a quick glance :)

**Comment by hmottestad** [comment]
@jeenbroekstra This fix here is impacting one of the SHACL tests, 3-4x slower than before. TruncatedValidationReportTest.

Not sure what's going on, just wanted to let you know.
